### PR TITLE
ref(toolbar): check allowed origins in iframe view

### LINKED ISSUE #77214: [DevToolbar] Guard `/iframe/` & `/login-success/` pages against disallowed referer origins
The `/toolbar/:org/:project/iframe/` view (not `/login-success/`) page should check that the `referer` header matches with the list of saved "Allowed Origins" within the project settings. This happens after the user is logged in.

We have the `project` url parameter, so we should be able to lookup the allow list. Project name/slug/id should all be supported in the url. Wildcard subdomains are possible in the allowlist, the port must match too if specified.

If the domain is allowed then we'll pass a variable into the template to indicate that. `domain_is_allowed = True`
Else, if the domain is not allowed, then the variable will be set to something else. `domain_is_allowed = True`

The specific variable doesn't super matter, what matters is:
- we're doing the check
- can see the result of the check in the template

In the future we might have two templates, one for each case. But at this point there's no need.

Depends on https://github.com/getsentry/sentry/issues/77213

### PR DESCRIPTION
**Pull request by aliu39**
Updates iframe view to lookup the project from the URL param, _after_ auth. Handles missing project and allowed origins for the request (set up by a project setting).

Decided with @ryan953 we'll only allow wildcards in the project settings, for the leftmost subdomain only.

Closes https://github.com/getsentry/sentry/issues/77214

### CODE REVIEW

**Comment by aliu39** [comment]
Ready, but needs test coverage:
- unit tests for `url_matches`
- e2e tests making requests to `iframe_view`
  - project doesn't exist
  - missing REFERER
  - same cases as `url_matches`

File: src/sentry/toolbar/views/iframe_view.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+@region_silo_view
+class IframeView(OrganizationView):
+    security_headers = {
+        "X-Frame-Options": "ALLOWALL"
+    }  # allows response to be embedded in an iframe.
+
+    def respond(self, template: str, context: dict[str, Any] | None = None, status: int = 200):
+        response = super().respond(template, context=context, status=status)
+        for header, val in IframeView.security_headers.items():
+            response[header] = val
```
Comment (cmanallen) [inline src/sentry/toolbar/views/iframe_view.py:23]: Let's just assign these statically.  I.e. without the class variable.  `response["X-Frame-Options"] = "ALLOWALL"`

**Review by cmanallen — CHANGES_REQUESTED** [review]
Finish the todos.  Add test coverage.

File: src/sentry/toolbar/utils/url.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    * port: optional for `url`. Must equal target.port, unless it is excluded from target.
+    * scheme: must be non-empty and equal target.scheme, unless it is excluded from target.
+    """
+
+    # We expect hostname exists for both urls.
+    hostname, target_hostname = url.hostname, target_url.hostname
+    if hostname.startswith("*."):
+        hostname = hostname.split(".", 1)[1]
+        if "." not in target_hostname:
+            return False
```
[the remark this answers is not in the document]
Comment (aliu39) [reply src/sentry/toolbar/utils/url.py:21]: are you saying "." will always be in a hostname?

[the remark this answers is not in the document]
Comment (aliu39) [reply src/sentry/toolbar/utils/url.py]: Nvm, I totally messed up the logic here. We allow wildcards in target, not the referrer.

**Comment by aliu39** [comment]
Do we have any need for this response header? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

**Comment by ryan953** [comment]
> Do we have any need for this response header? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

Maybe! But that'll be a separate PR i think. If we were to use that we would still want to run the check in python first to verify which type of response to return. If the domain is valid we might want to add that header with only the valid domain listed... I would not want to return a list of valid domains because that could be a leak of customer info.

ticketed: https://github.com/getsentry/sentry-toolbar/issues/29

**Review by ryan953 — APPROVED** [review]
looks good to me. I can't think of missing test cases right now

**Review by cmanallen — APPROVED** [review]

**Pull request merged**